File: backend/src/routes/chatbot.js
import { Router } from 'express';
import db from '../db.js';
import { authenticate } from '../middleware/auth.js';

const router = Router();
router.use(authenticate);

const conversations = new Map();

router.post('/message', async (req, res) => {
  const { message } = req.body;
  if (!message) return res.status(400).json({ error: 'Mensaje requerido' });

  if (!conversations.has(req.userId)) {
    conversations.set(req.userId, []);
  }
  const history = conversations.get(req.userId);
  history.push({ role: 'user', content: message });

  const user = db.prepare('SELECT name FROM users WHERE id = ?').get(req.userId);
  const summary = db.prepare(`
    SELECT COALESCE(SUM(CASE WHEN type = 'income' THEN amount ELSE 0 END), 0) as total_income,
      COALESCE(SUM(CASE WHEN type = 'expense' THEN amount ELSE 0 END), 0) as total_expense
    FROM transactions WHERE user_id = ?
  `).get(req.userId);

  const recentTx = db.prepare(`
    SELECT t.date, t.type, t.amount, t.description, c.name as category
    FROM transactions t JOIN categories c ON t.category_id = c.id
    WHERE t.user_id = ? ORDER BY t.date DESC LIMIT 5
  `).all(req.userId);

  const goals = db.prepare('SELECT name, target_amount, current_amount FROM savings_goals WHERE user_id = ?').all(req.userId);

  const systemPrompt = `Eres un asesor financiero personal. Ayudas al usuario con sus finanzas personales.
Usuario: ${user?.name || 'Usuario'}
Resumen financiero:
- Ingresos totales: $${summary?.total_income || 0}
- Gastos totales: $${summary?.total_expense || 0}
- Balance: $${((summary?.total_income || 0) - (summary?.total_expense || 0)).toFixed(2)}

Últimas transacciones:
${recentTx.map(t => `- ${t.date}: ${t.type === 'income' ? 'Ingreso' : 'Gasto'} de $${t.amount} en ${t.category}${t.description ? ': ' + t.description : ''}`).join('\n')}

Metas de ahorro:
${goals.map(g => `- ${g.name}: $${g.current_amount}/$${g.target_amount} (${Math.round(g.current_amount / g.target_amount * 100)}%)`).join('\n') || 'Ninguna'}

Responde de forma útil, clara y concisa. Si te preguntan algo que no sabes, sé honesto.`;

  const fullPrompt = `${systemPrompt}\n\nHistorial:\n${history.slice(-6).map(m => `${m.role === 'user' ? 'Usuario' : 'Asistente'}: ${m.content}`).join('\n')}\n\nAsistente:`;

  try {
    const response = await queryOllama(fullPrompt);
    history.push({ role: 'assistant', content: response });
    if (history.length > 20) history.splice(0, history.length - 20);
    res.json({ response });
  } catch (err) {
    res.status(500).json({ error: 'Error al contactar el asistente: ' + err.message });
  }
});

router.post('/clear', (req, res) => {
  conversations.delete(req.userId);
  res.json({ message: 'Conversación reiniciada' });
});

async function queryOllama(prompt) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 30000);

  try {
    const OLLAMA_URL = process.env.OLLAMA_URL || 'http://localhost:11434';
    const resp = await fetch(`${OLLAMA_URL}/api/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: 'llama3.2',
        prompt,
        stream: false,
        options: { temperature: 0.7, max_tokens: 500 }
      }),
      signal: controller.signal
    });

    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Ollama error: ${resp.status} - ${text}`);
    }

    const data = await resp.json();
    return data.response.trim();
  } finally {
    clearTimeout(timeout);
  }
}

export default router;

File: ai/chatbot.py
import os
import threading
import time

import db
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from rag import indexer
from rag.embeddings import get_embeddings
from rag.llm import get_chat_model
from rag.retriever import Retriever
from rag.vector_store import ChromaVectorStoreProvider

INDEX_TTL = int(os.environ.get("INDEX_TTL", "300"))


class ChatService:
    def __init__(self):
        embeddings = get_embeddings()
        self._llm = get_chat_model()
        self._provider = ChromaVectorStoreProvider(embeddings)
        self._kb_store = None
        self._lock = threading.Lock()
        self._indexed_at = {}
        self._conversations = {}

    def process_message(self, user_id, user, message):
        self._ensure_kb()
        self._ensure_user_index(user_id)

        aggregates = db.get_aggregates(user_id)
        balance = aggregates["total_income"] - aggregates["total_expense"]

        context = Retriever(self._provider, self._kb_store).retrieve_context(user_id, message)

        system_prompt = f"""Eres un asesor financiero personal. Ayudas al usuario con sus finanzas personales.
Usuario: {user['name']}

Resumen financiero:
- Ingresos totales: ${aggregates['total_income']:,.0f}
- Gastos totales: ${aggregates['total_expense']:,.0f}
- Balance: ${balance:,.0f}

Información recuperada (RAG) para responder con precisión:
{context}

Responde de forma útil, clara y concisa, en español. Usa los datos recuperados para responder; si no sabes algo, sé honesto."""

        history = self._conversations.setdefault(user_id, [])
        messages = [SystemMessage(content=system_prompt)]
        for m in history[-12:]:
            if m["role"] == "user":
                messages.append(HumanMessage(content=m["content"]))
            else:
                messages.append(AIMessage(content=m["content"]))
        messages.append(HumanMessage(content=message))

        response = self._llm.invoke(messages)
        answer = response.content if isinstance(response.content, str) else str(response.content)

        history.append({"role": "user", "content": message})
        history.append({"role": "assistant", "content": answer})
        if len(history) > 20:
            del history[: len(history) - 20]

        return answer

    def clear(self, user_id):
        self._conversations.pop(user_id, None)

    def _ensure_kb(self):
        if self._kb_store is None:
            with self._lock:
                if self._kb_store is None:
                    self._kb_store = self._provider.create_kb_store(indexer.build_kb_documents())

    def _ensure_user_index(self, user_id):
        now = time.time()
        with self._lock:
            if self._indexed_at.get(user_id, 0) + INDEX_TTL > now:
                return
            self._provider.upsert(user_id, indexer.build_user_documents(user_id))
            self._indexed_at[user_id] = now